core/src/Http/HttpRequest.cs

```diff
@@ -74,16 +80,39 @@ private static HttpRequestMessage CreateRequestMessage(Uri endpoint, Dictionary<
             return requestMessage;
         }
 
+        private static async Task<HttpContent> CloneContent(HttpContent content)
+        {
+            var temp = new System.IO.MemoryStream();
+            await content.CopyToAsync(temp).ConfigureAwait(false);
+            temp.Position = 0;
+            var clone = new StreamContent(temp);
+            if (content.Headers != null)
+            {
+                foreach (var h in content.Headers)
+                {
+                    clone.Headers.Add(h.Key, h.Value);
+                }
+            }
+            return clone;
+        }
+
         private static async Task<HttpResponse> ExecuteWithRetry(Uri endpoint, Dictionary<string, string> headers,
-            Dictionary<string, string> bodyParameters, HttpMethod method,
+            HttpContent body, HttpMethod method,
             RequestContext requestContext, bool retry = true)
         {
             Exception toThrow = null;
             bool isRetryable = false;
             HttpResponse response = null;
             try
             {
-                response = await Execute(endpoint, headers, bodyParameters, method);
+                HttpContent content = null;
+                if(body != null)
+                {
+                    // Since HttpContent would be disposed by underlying client.SendAsync(),
+                    // we duplicate it so that we will have a copy in case we would need to retry
```

Comment:
comment not needed here imo. if really needed should rather be an xml comment on the private method... consider though naming the method signaling the intent and reasoning.

Reply:
Copied from earlier reply:

I have a different perspective. In general, most typically-short function names would signal the intent (i.e. what does it do), but hardly long enough to explain the reasoning (i.e. why you would call this function but not its siblings). And the reasoning is probably determined by the caller side, not by the callee.

In this particular case, the reason to have a deep copy is explained in those 2-line comment. This was not obvious and, in fact, I did not notice this subtle issue at the beginning, until later I needed time to debug and added a dedicated commit to fix it. So perhaps those comments is worth something here.


---
In reply to: [201542977](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1118#discussion_r201542977) [](ancestors = 201542977)